Refactor SpotData CLI to change usage of forecast times

gavinevans opened the pull request:
Issue: #217 

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

gavinevans commented on bin/improver-spot-extract:
```diff
@@ -247,37 +239,37 @@ def main():
-        diagnostics[diagnostic_key]["data"] = Load('multi_file').process(
-            files_to_read, diagnostic_name)
+        diagnostics[diagnostic_key]["data"] = (
+            Load('multi_file').process(files_to_read))
```
I think the reason for taking this out is because the name of the diagnostic within the file doesn't always match the `diagnostic_name`. This can be imagined for probabilistic cubes, where currently the input cube may have a name, such as `probability_of_air_temperature`, rather than `air_temperature`, so loading when constraining upon `diagnostic_name` will return no cubes.

gavinevans commented on bin/improver-spot-extract:
```diff
@@ -247,37 +239,37 @@ def main():
             use_multiprocessing=args.multiprocess))
 
-    for resulting_cube in resulting_cubes:
+    for resulting_cube, filename in zip(resulting_cubes, all_available_files):
```
I've reinstated this concatenation behaviour. Currently due to the integration into the suite, this behaviour isn't being used, however, I've tested using the CLI test where the input cube has two cubes, and the concatenation over time seemed to work in that case.

gavinevans commented on lib/improver/spotdata/main.py:
```diff
@@ -217,15 +196,19 @@ def run_spotdata(diagnostics, ancillary_data, sites, config_constants,
-            resulting_cubes.append(resulting_cubelist)
-            extrema_cubes.append(extrema_cubelist)
+            if not resulting_cubelist:
+                resulting_cubelist = [resulting_cubelist]
```
I've removed this, as I don't think it's necessary.

gavinevans commented on bin/improver-spot-extract:
```diff
@@ -226,17 +221,14 @@ def main():
-
     # Create a list of files that will need to be read.
-    all_available_files = glob.glob(args.data_path+"/*")
+    all_available_files = glob.glob(args.data_path)
```
I've left this as it is, and modified the docstring of the command line argument to indicate that `args.data_path` is now expected as a file path i.e. `/path/to/file.nc`, rather than `/path/to/`. This fits with the current usage of SpotData, however, this may need to be revisited, once we're clearer on the usage of SpotData.

bayliffe commented on bin/improver-spot-extract:
```diff
@@ -245,36 +236,41 @@ def main():
-        if resulting_cube:
-            WriteOutput(
-                'as_netcdf', dir_path=args.output_path).process(resulting_cube)
+    filename = os.path.splitext(os.path.basename(all_available_files[0]))[0]
```
NB: This won't work when taking multiple files (we may get a different diagnostic name to that being processed).

bayliffe commented on bin/improver-spot-extract:
```diff
@@ -245,36 +236,41 @@ def main():
+    if resulting_cubes:
+        # Concatenate CubeList into Cube, creating a time DimCoord,
+        # and write out.
+        cube_out, = resulting_cubes.concatenate()
```
NB: This won't work when processing multiple diagnostics. Both this and the saving will need to be moved within the diagnostics loop above.

LaurenceBeard commented on lib/improver/spotdata/site_data.py:
```diff
@@ -103,7 +103,7 @@ def from_file(self, file_path):
         if file_path.endswith('.json'):
             data_file = open(file_path, 'r')
-            site_data = json.load(data_file)
+            site_data = json.load(data_file)["sites"]
```
Are we expecting something else to be in the site_data file, other than a list of sites?

If you remove:
```
{
	"sites": 
```
from the start of the json, and:
```
}
```
from the end, the json still validates using https://jsonlint.com/

gavinevans replied:
I've removed the usage of a dictionary within the json file, so that a list is used instead.

carolinejjones commented on tests/improver-spot-extract/02-nearest.bats:
```diff
@@ -39,24 +39,21 @@
       --diagnostics temperature \
       --latitudes 10 20 30 40 50 60 \
       --longitudes 0 0 0 0 0 0 \
       --altitudes 0 1 2 3 4 5 \
```
I think you need to remove the \ to get this to work

gavinevans replied:
Removed.

gavinevans commented:
@carolinejjones , I think I've corrected the problems with the CLI tests now, and I've moved the test data centrally, so that you can run the CLI tests without pointing to my local copy of the CLI test data. The spot-extract CLI tests now run successfully for me, as I've removed the trailing slash as you've recommended, and I've corrected the names of the files that are being removed at the end of the CLI, as previously these names were incorrect and causing the CLI tests to fail. The only changes that I've made to the test data is to rename orography.nc to highres_orog.nc, and to update nearest_diagnostics.json and model_level_diagnostics.json.

carolinejjones commented:
All the tests now run through correctly.
